flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -0,0 +1,268 @@
+			ExecutionSlotSharingGroup executionSlotSharingGroup,
+			Collection<ExecutionVertexID> executions) {
+		SharedSlot sharedSlot = getOrAllocateSharedSlot(executionSlotSharingGroup, sharedSlotProfileRetriever);
+		return executions.stream().map(sharedSlot::createAssignment);
+	}
+
+	private SharedSlot getOrAllocateSharedSlot(
+			ExecutionSlotSharingGroup executionSlotSharingGroup,
+			SharedSlotProfileRetriever sharedSlotProfileRetriever) {
+		return sharedSlots
+			.computeIfAbsent(executionSlotSharingGroup, group -> {
+				SlotRequestId physicalSlotRequestId = new SlotRequestId();
+				CompletableFuture<PhysicalSlot> physicalSlotFuture = sharedSlotProfileRetriever
+					.getSlotProfileFuture(group)
+					.thenCompose(slotProfile -> slotProvider.allocatePhysicalSlot(
+						new PhysicalSlotRequest(physicalSlotRequestId, slotProfile, slotWillBeOccupiedIndefinitely)))
+					.thenApply(PhysicalSlotRequest.Result::getPhysicalSlot);
+				return new SharedSlot(physicalSlotRequestId, group, physicalSlotFuture);
+			});
+	}
+
+	private class SharedSlot implements SlotOwner, PhysicalSlot.Payload {
+		private final SlotRequestId physicalSlotRequestId;
+
+		private final ExecutionSlotSharingGroup executionSlotSharingGroup;
+
+		private final CompletableFuture<PhysicalSlot> slotContextFuture;
+
+		private final DualKeyLinkedMap<ExecutionVertexID, SlotRequestId, SlotExecutionVertexAssignment> requestedLogicalSlots;
+
+		private SharedSlot(
+				SlotRequestId physicalSlotRequestId,
+				ExecutionSlotSharingGroup executionSlotSharingGroup,
+				CompletableFuture<PhysicalSlot> slotContextFuture) {
+			this.physicalSlotRequestId = physicalSlotRequestId;
+			this.executionSlotSharingGroup = executionSlotSharingGroup;
+			this.slotContextFuture = slotContextFuture.thenApply(physicalSlot -> {
+				Preconditions.checkState(
+					physicalSlot.tryAssignPayload(this),
+					"Unexpected physical slot payload assignment failure!");
+				return physicalSlot;
+			});
+			this.requestedLogicalSlots = new DualKeyLinkedMap<>(executionSlotSharingGroup.getExecutionVertexIds().size());
+		}
+
+		SlotExecutionVertexAssignment createAssignment(ExecutionVertexID executionVertexId) {
+			Preconditions.checkArgument(executionSlotSharingGroup.getExecutionVertexIds().contains(executionVertexId));
+			SlotRequestId logicalSlotRequestId = new SlotRequestId();
+			CompletableFuture<LogicalSlot> logicalSlotFuture = allocateLogicalSlot(logicalSlotRequestId, executionVertexId);
+			SlotExecutionVertexAssignment assignment = new SlotExecutionVertexAssignment(executionVertexId, logicalSlotFuture);
+			requestedLogicalSlots.put(
+				executionVertexId,
+				logicalSlotRequestId,
+				assignment);
+			return assignment;
+		}
+
+		private CompletableFuture<LogicalSlot> allocateLogicalSlot(
+				SlotRequestId logicalSlotRequestId,
+				ExecutionVertexID executionVertexId) {
+			String logMessageBase = getLogicalSlotString(logicalSlotRequestId, executionVertexId);
```

Comment:
Maybe add a sanity check here to ensure that the `slotContextFuture` is not completed exceptionally here?
Otherwise the failed slot future will be registered to `requestedLogicalSlots` cannot be cleared.

Reply:
I will change the order:
```
requestedLogicalSlots.put(executionVertexId, logicalSlotRequestId, logicalSlotFuture);
logicalSlotFuture.exceptionally(cause -> {
	LOG.debug("Failed {}", logMessageBase);
	cancelLogicalSlotRequest(logicalSlotRequestId);
	return null;
});			
```
In this case, if `slotContextFuture` has already completed exceptionally, it will fail the `logicalSlotFuture` immediately and `logicalSlotFuture.exceptionally -> cancelLogicalSlotRequest` should do the cleanup.